File: burgers1d/python/src/main_fom_velocity.py

import numpy as np
import sys, time
import scipy.linalg.blas as scblas
# local app class
from burgers1d import Burgers1d

meshSize = int(sys.argv[1])

np.set_printoptions(linewidth=400)
print(meshSize)
# create app
appObj = Burgers1d(meshSize)
yRef = np.ones(meshSize)
t0 = 0.0

romSize = 100
numRound = 5000

# phi_C means that phi_C.T will be col major
phi_C = np.ones((meshSize, romSize), order = 'C')
print (phi_C.flags)
phi_F = np.ones((meshSize, romSize), order = 'F')
print (phi_F.flags)

phi = phi_F
a = np.zeros(romSize)
a1 = np.zeros(meshSize)

#--- one run for warmup ---
f = appObj.velocity(yRef, t0)
a = np.dot(phi.T, f)
a1 = np.dot(phi, a)

#scblas.dgemv(1.0, phi, f, 0.0, A, 0, 1, 0, 1, 1, 1)
# J = np.zeros((meshSize, meshSize))
# appObj.jacobian(yRef, t0, J)

#--- do many replicas ---
startTime = time.time()
for i in range(numRound):
  appObj.velocity(yRef, t0, f)
  #scblas.dgemv(1.0, phi, f, 0.0, A, 0, 1, 0, 1, 1, 1)
  a  = np.dot(phi.T, f)
  a1 = np.dot(phi, a)

endTime = time.time()
elapsed = endTime-startTime
print("Elapsed time: {0:10.10f} ".format(elapsed) )
print("Time per round: {0:10.10f} ".format(elapsed/numRound) )

File: burgers1d/cpp/src/main_fom_velocity.cc
#define EIGEN_USE_BLAS
#define EIGEN_USE_LAPACKE

#include "CONTAINERS_ALL"
#include "ODE_ALL"
#include "burgers1d_eigen.hpp"
#include "burgers1d_input_parser.hpp"
#include "eigen_observer.hpp"

int main(int argc, char *argv[])
{
  // types
  using app_t		= Burgers1dEigen;
  using scalar_t	= typename app_t::scalar_type;
  using app_state_t	= typename app_t::state_type;
  using app_jacob_t	= typename app_t::jacobian_type;
  using app_dmat_t	= typename app_t::eig_dense_mat;
  using ode_state_t	= pressio::containers::Vector<app_state_t>;

  constexpr auto zero = ::pressio::utils::constants::zero<scalar_t>();
  constexpr auto one  = ::pressio::utils::constants::one<scalar_t>();

  // parse input file
  InputParser parser;
  int err = parser.parse(argc, argv);
  if (err == 1) return 1;

  // create app object
  app_t appObj(parser.numCell_);
  // set initial state to dummy, all ones
  app_state_t x0n(parser.numCell_);
  x0n.setConstant(one);

  // app_jacob_t J(parser.numCell_, parser.numCell_);
  // appObj.jacobian(x0n, zero, J);

  const int32_t numRounds = 5000;
  const int32_t romSize = 100;

  app_dmat_t phi(parser.numCell_, romSize);
  phi.setConstant(one);
  app_state_t a(romSize);
  app_state_t a1(parser.numCell_);

  //--- do one for warm up ---
  auto f = appObj.velocity(x0n, zero);
  a  = phi.transpose() * f;
  a1 = phi * a;

  //--- do many runs ---
  auto startTime = std::chrono::high_resolution_clock::now();
  for (auto i=0; i<numRounds; ++i){
    appObj.velocity(x0n, zero, f);
    a  = phi.transpose() * f;
    a1 = phi * a;
    //appObj.jacobian(x0n, zero, J);
  }

  auto finishTime = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed = finishTime - startTime;
  std::cout << "Elapsed time: "
  	    << std::fixed << std::setprecision(10)
  	    << elapsed.count() << std::endl;

  std::cout << "Time per round: "
  	    << std::fixed << std::setprecision(10)
  	    << elapsed.count()/(double)numRounds << std::endl;

  return 0;
}
